packages/apollo-gateway/src/datasources/RemoteGraphQLDataSource.ts

```diff
@@ -52,19 +72,77 @@ export class RemoteGraphQLDataSource implements GraphQLDataSource {
       await this.willSendRequest({ request, context });
     }
 
-    const { http, ...graphqlRequest } = request;
-    const options: RequestInit = {
-      ...http,
-      body: JSON.stringify(graphqlRequest),
-    };
+    if (!request.query) {
+      throw new Error("Missing query");
+    }
+
+    const apqHash = createSHA('sha256')
+       .update(request.query)
+       .digest('hex');
 
-    const httpRequest = new Request(request.http.url, options);
+    const { query, ...requestWithoutQuery } = request;
 
     const respond = (response: GraphQLResponse, request: GraphQLRequest) =>
       typeof this.didReceiveResponse === "function"
         ? this.didReceiveResponse({ response, request, context })
         : response;
 
+    if (this.apq) {
+      // Take the original extensions and extend them with
+      // the necessary "extensions" for APQ handshaking.
+      requestWithoutQuery.extensions = {
+        ...request.extensions,
+        persistedQuery: {
+          version: 1,
+          sha256Hash: apqHash,
+        },
+      };
+
+      const apqOptimisticResponse =
+        await this.sendRequest(requestWithoutQuery, context);
+
+      // If we didn't receive notice to retry with APQ, then let's
+      // assume this is the best result we'll get and return it!
+      if (
+        !apqOptimisticResponse.errors ||
+        !apqOptimisticResponse.errors.find(error =>
+          error.message === 'PersistedQueryNotFound')
+      ) {
+        return respond(apqOptimisticResponse, requestWithoutQuery);
+      }
+    }
+
+    // If APQ was enabled, we'll run the same request again, but add in the
+    // previously omitted `query`.  If APQ was NOT enabled, this is the first
+    // request (non-APQ, all the way).
+    const requestWithQuery: GraphQLRequest = {
+      query,
+      ...requestWithoutQuery,
+    };
```

Comment:
Minor nit: I think `requestWithQuery === request` from above (the one that we destructured).

Reply:
I'm wrong, since in the APQ case, `requestWithoutQuery.extensions` is mutated and it seems we need to pass those along in the request as well. I'll leave this for you to resolve in case you arrive at a different conclusion than I.